### PR TITLE
Show feedback prompt only once

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR resolves the problem described in [this issue](https://github.com/woocommerce/woocommerce/issues/39167). We want to show the feedback prompts (to add and edit a product)only once in the legacy product experience.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a WooCommerce store with this branch and (**don't enable** the new product editor).
2. Under **WooCommerce > Settings > Advanced > Woo.com** verify that the `Enable tracking` option is `enabled`.  
3. Go to **Products > Add New**.
4. Add a product name and press `Publish`.
5. Verify the feedback prompt with the text `How easy was it to add a product?` is shown.

Screenshot 2023-12-28 at 16 35 24

6. Edit that product and verify the feedback prompt with the text `How easy was it to edit your product?` is visible.

Screenshot 2023-12-28 at 16 34 21

7. Add a change and save.
9. Verify that the feedback prompt is not being shown.
10. Repeat steps 3 and 4.
11. The feedback prompt shouldn't be shown.
12. If you want to test this again, you should set the option: `woocommerce_ces_shown_for_actions` to `a:0:{}` by using the WCA Test Helper under Tools. 

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js [hunk 1]
```diff
@@ -58,7 +58,15 @@ function _CustomerEffortScoreTracks( {
 	updateOptions,
 	createNotice,
 } ) {
-	const [ modalShown, setModalShown ] = useState( false );
+	useEffect( () => {
+		if ( cesShownForActions && ! cesShownForActions.includes( action ) )
```
Comment (louwie17) [inline packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js]: This logic is wrong, it should be if `cesShownForActions` does not exist or `! cesShownForActions.includes( action )`
Also calling it on initial render seems to break this, as the component is not done `resolving` its data yet, so `cesShownForActions` is always `undefined`, meaning that the previous value will always be overwritten.

I think a better option would be to add this logic in the `onNoticeShown` callback, this way we always know the actual component has finished resolving and tracking is also enabled.
We can still add the `if ( ! cesShownForActions || ! cesShownForActions.includes( action ) )` check there to make sure we don't add duplicates.

**Review by louwie17 — CHANGES_REQUESTED** [review]
@octaedro I noticed a couple bugs, and suggested some changes inline. I am actually not a 100% sure how this was working for you?

File: packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js (hunk 1 above)
Comment (octaedro) [reply packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js]: Thank you @louwie17 for reviewing this.
I forgot to push my latest changes 🤦🏻‍♂️. Apologies for that 🙏
I fixed it in the commit [078074e](https://github.com/woocommerce/woocommerce/pull/43164/commits/078074ed68b4e93510033b78533e8e4eae099ff1)

File: packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js [hunk 2]
```diff
@@ -69,37 +67,21 @@ function _CustomerEffortScoreTracks( {
...
-	// (we don't want to return null early), if the modal was shown for this
-	// instantiation, so that the component doesn't go away while we are
-	// still showing it.
-	if (
-		cesShownForActions &&
-		cesShownForActions.indexOf( action ) !== -1 &&
-		! modalShown
-	) {
-		return null;
-	}
```
Comment (louwie17) [inline packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js]: I do think we want to keep this condition here, I am not a 100% sure about the `! modalShown` logic, that could potentially be removed, I don't quite understand that part of the logic. So maybe its worth to also keep that.

Reason for needing this is that the ces could still be triggered from the `client` in which case we did have to check `cesShownForActions` first. One way to test this is the `product_add_view` action, by doing:
1.  Go to **Products > Add new**
2. Add some changes without saving
3. Click **WooCommerce > Home**
4. Click **Leave** when it prompts you about the unsaved changes
5. Notice the `notice` that shows up, currently this notice shows up every time, adding the above logic will cause it to only show once.

BTW I found this by git blaming the above code and looking at the PR that added it, I guess I had added/updated it as part of this PR: https://github.com/woocommerce/woocommerce/pull/35728 ( which includes the above scenario ).

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for the updates @octaedro, it tested well, but did notice one issue around some of the removed logic that we should probably keep.

File: packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js (hunk 2 above)
Comment (octaedro) [reply packages/js/customer-effort-score/src/components/customer-effort-score-tracks/index.js]: Good eyes @louwie17! As you mentioned, we needed an early `null` return to not show the prompt again.

**Review by louwie17 — APPROVED** [review]
LGTM 🚀 Thanks for sticking with the updates @octaedro, this looks great and tested well!

**Pull request merged**